oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/OptimizerEndPoint.kt

```diff
@@ -0,0 +1,284 @@
+                                val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                                        OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(simName)
+                                ).build()
+                                sim.input.onNext(message)
+                                select {
+                                    sim.output.onReceive { SimResult.TimeOut(it.name) }
+                                    sim.error.onReceive { SimResult.TimeOutFailure(it.name, it.exception) }
+                                }
+                            }
+                        }
+                    }
+
+                    val newResultMessage: Message
+                    val resultMessage: Result
+                    when (result) {
+                        is SimResult.Success -> {
+                            newResultMessage = Message(result.name, "Evaluation Succeed: Result [${result.result}]")
+                            resultMessage = Result(result.name, "Success", inputVector.toString(), result.result.toString())
+                        }
+                        is SimResult.Failure -> {
+                            newResultMessage = Message(result.name, "Evaluation Failed: Due to\n${result.exception}")
+                            resultMessage = Result(result.name, "Error", inputVector.toString(), "Failure: \n${result.exception}")
+                        }
+                        is SimResult.TimeOut -> {
+                            newResultMessage = Message(result.name, "Evaluation Canceled to timed out after ${sim.timeOut!!.toMillis()} millisecond")
+                            resultMessage = Result(result.name, "Canceled", inputVector.toString(), "Timeout")
+                        }
+                        is SimResult.TimeOutFailure -> {
+                            newResultMessage = Message(result.name, "Evaluation try to cancel Timed out after ${sim.timeOut!!.toMillis()} millisecond but Failed: Due to\n${result.exception}")
+                            resultMessage = Result(result.name, "Cancel Failed", inputVector.toString(), "Cancellation failed:\n${result.exception}")
+                        }
+                    }
+                    modelService.addMessage(newResultMessage)
+                    modelService.addResult(resultMessage)
+                    toolNumber++
+                }
+            }
+            modelService.updateStatusMessage("Idle")
+        }
+    }
+
+
+    override fun stopOptimization() {
+        state = State.Idle
+    }
+
+    override fun syncAll() {
+        GlobalScope.launch {
+            for (simName in modelService.simByName.keys) {
+                syncConfigFor(simName)
+            }
+        }
+    }
+
+    override fun updateNode(simName: String) {
+        GlobalScope.launch {
+            syncConfigFor(simName)
+        }
+    }
+
+    override fun cancelAll() {
+        GlobalScope.launch {
```

Comment:
I suspect you havent written any tests on this class.

I know this is true because if you tried to test `cancelALl` you'd have no way of synchronizing on the result.

This code makes `cancelAll` not just parallel by default, but _impossible_ to make serial. I dont think this is a good idea.

Reply:
I change all the call to suspend